### PR TITLE
Code snippet dropdowns

### LINKED ISSUE #3469: Implement dropdown menus for code snippet

### PR DESCRIPTION
**Pull request by sowasred2012**
## Done

Added example of using a select within a code snippet

Fixes #3469 

## QA

- Open [demo](https://vanilla-framework-3511.demos.haus/docs/examples/patterns/code-snippet/code-snippet-dropdown)
- Check that the example works when selecting different options from the menu
- Compare to the [Zeplin design](https://app.zeplin.io/project/5fa270d8eee4dbb267e579f1/screen/5ffc2fdcb65e6a99b78f9d47)
- Review updated documentation:
  - [Code](https://vanilla-framework-3511.demos.haus/docs/base/code)
  - [Component status](https://vanilla-framework-3511.demos.haus/docs/component-status)

### CODE REVIEW

**Comment by bartaz** [comment]
On Zeplin it seems that we should have different background on the button when menu is open:
Screenshot 2021-01-26 at 07 53 59

I don't think we have a 'pressed' state on buttons ATM, so not sure if that's something we should add custom to code snippet button, or can it wait for when buttons have pressed state in Vanilla @lyubomir-popov ?

File: templates/docs/examples/patterns/code-snippet/_script.js [hunk 1]
```diff
@@ -0,0 +1,34 @@
+function setupCodeSnippetOptions(dropdown) {
+  var dropdownLabel = dropdown.querySelector('span');
+  var targetMenu = document.getElementById(dropdown.getAttribute('aria-controls'));
+  var options = [].slice.call(targetMenu.querySelectorAll('a'));
```
Comment (bartaz) [inline templates/docs/examples/patterns/code-snippet/_script.js:4]: We shouldn't target options via tag name. Using class name would be better (we know it as it's part of the pattern), but I guess even better we could check for aria-controls attribute maybe? `[aria-controls]`? Because we rely on this attribute for the functionality anyway.

Comment (sowasred2012) [reply templates/docs/examples/patterns/code-snippet/_script.js]: good point, updated

**Comment by sowasred2012** [comment]
> On Zeplin it seems that we should have different background on the button when menu is open:
> Screenshot 2021-01-26 at 07 53 59
> 
> I don't think we have a 'pressed' state on buttons ATM, so not sure if that's something we should add custom to code snippet button, or can it wait for when buttons have pressed state in Vanilla @lyubomir-popov ?

Agreed in standup to address in a separate PR that covers buttons in general.

**Comment by sowasred2012** [comment]
After a debate - decided to go with `select`s for the dropdowns, rather than contextual menus.

File: scss/_patterns_code-snippet.scss [hunk 2]
```diff
@@ -63,5 +63,25 @@ $color-snippet-heading-bg: rgba($color-x-dark, 0.08);
...
+      background-color: $color-transparent;
+      border-color: $color-transparent;
+      color: $color-dark;
+
+      // taken from the .u-table-cell-padding-overlap utility
+      margin-bottom: -#{$table-cell-vertical-padding} !important;
+      margin-top: calc(-#{$table-cell-vertical-padding} - 1px) !important;
+      min-width: min-content;
+
+      // taken from the .is-dense button modifier
```
Comment (bartaz) [inline scss/_patterns_code-snippet.scss:82]: Does this still apply when it's a select rather than a button?

File: templates/docs/component-status.md [hunk 3]
```diff
@@ -23,13 +23,20 @@ When we add, make significant updates, or deprecate a component we update their
...
+      <td>We add the ability to include a select within in a code snippet, allowing users to switch between related code examples.</td>
+    </tr>
     <tr>
       <th><a href="/docs/base/code#code-snippet">Code snippet</a></th>
       <td><div class="p-label--updated">Updated</div></td>
       <td>2.23.0</td>
       <td>We added a utility class for <code>.p-code-snippet__block</code> that allows content to wrap, rather than horizontally scroll: <code>.is-wrapped</code>.</td>
     </tr>
   </tbody>
+  </tbody>
```
Comment (bartaz) [inline templates/docs/component-status.md:39]:
```suggestion
```

Duplicated tbody?

File: templates/docs/examples/patterns/code-snippet/_script.js [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+    var targetElement = document.getElementById(e.target.value);
+
+    togglePanel(targetElement, dropdown.options);
+  });
+}
+
+function togglePanel(targetPanel, options) {
+  for (var i = 0; i < options.length; i++) {
+    var panel = document.getElementById(options[i].value);
+    panel.classList.add('u-hide');
```
Comment (bartaz) [inline templates/docs/examples/patterns/code-snippet/_script.js:12]: Should we be toggling aria-hidden as well?

File: templates/docs/examples/patterns/code-snippet/_script.js [hunk 5]
```diff
@@ -0,0 +1,26 @@
+function attachDropdownEvents(dropdown) {
```
Comment (bartaz) [inline templates/docs/examples/patterns/code-snippet/_script.js:1]: I think we tend to add a bit of a jsdoc comment on top of those in other examples, to give better idea of what the function does and what the params mean.
Could you have a look at some other example JS and use similar comment style?

File: templates/docs/base/code.md [hunk 6]
```diff
@@ -60,6 +60,14 @@ View example of the code snippet
 View example of the code snippet
 </a></div>
 
+**Code snippet with dropdown**
```
Comment (bartaz) [inline templates/docs/base/code.md:63]: Now, that these are not  really 'dropdowns', does it still make sense to call them like this?

I wouldn't call them 'selects' (as that's HTML jargon), but what are they by function? Dropdowns? Menus? How does the spec call them?

File: templates/docs/examples/patterns/code-snippet/code-snippet-dropdown.html [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+
+  <pre id="html-panel" class="p-code-snippet__block--numbered is-wrapped"><code>&#60;h1 class="p-heading--two"&#62;How to use code snippets&#60;/h1&#62;</code></pre>
+
+  <pre id="css-panel" class="p-code-snippet__block u-hide"><code>.p-heading--two {
+color: red;
+}</code></pre>
+
+  <pre id="js-panel" class="p-code-snippet__block u-hide"><code>console.log("Example 1");</code></pre>
+
+  <div class="p-code-snippet__header">
```
Comment (bartaz) [inline templates/docs/examples/patterns/code-snippet/code-snippet-dropdown.html:28]: Do we need the Output part? It doesn't seem to be relevant to the dropdown, so maybe we can drop it to simplify things.

**Comment by bartaz** [comment]
Wow, switching to select made things so much better in terms of accessibility, simplicity of the code, etc. It was definitely a good call.

File: scss/_patterns_code-snippet.scss (hunk 2 above)
Comment (sowasred2012) [reply scss/_patterns_code-snippet.scss:82]: Yeah, I initially removed these when making the switch to selects, but there was way too much space around the element. Adding these back in solved that.

File: templates/docs/examples/patterns/code-snippet/code-snippet-dropdown.html (hunk 7 above)
Comment (sowasred2012) [reply templates/docs/examples/patterns/code-snippet/code-snippet-dropdown.html:28]: Not _really_, but it is the output you'd get from the JS? :man_shrugging:

File: templates/docs/base/code.md (hunk 6 above)
Comment (sowasred2012) [reply templates/docs/base/code.md:63]: [Here](https://discourse.ubuntu.com/t/code-snippet-pattern/19195) they're called "version switchers", in [the design](https://app.zeplin.io/project/5fa270d8eee4dbb267e579f1/screen/5ffc2fdcb65e6a99b78f9d47) they're called "dropdowns".

I did initially go with just calling them selects, but it felt inappropriate and like it was tied to a form. "Version switcher" won't suit every use case, but we could also go with "switcher", it's ultimately what the element does. That said, I think they can also be considered as "dropdowns", as it's _also_ what the element does. "Dropdown switcher"?

File: templates/docs/examples/patterns/code-snippet/_script.js (hunk 4 above)
Comment (sowasred2012) [reply templates/docs/examples/patterns/code-snippet/_script.js:12]: Good point

File: templates/docs/examples/patterns/code-snippet/code-snippet-dropdown.html (hunk 7 above)
Comment (bartaz) [reply templates/docs/examples/patterns/code-snippet/code-snippet-dropdown.html:28]: I find it bit confusing, as it's shown regardless of the lang chosen in the dropdown - it doesn't change, is not affected by the dropdown.

I feel that it would be better to get only what's needed into the example.

But, not a big deal. I don't mind if it says. Just something to consider.

**Review by bartaz — APPROVED** [review]
LGMT, thanks.
